cloudsim-plus/src/main/java/org/cloudbus/cloudsim/brokers/DatacenterBrokerAbstract.java

```diff
@@ -771,14 +774,71 @@ public void requestIdleVmDestruction(final Vm vm) {
         }
     }
 
+    private Cloudlet extractCloudlet(CloudletExecution cloudletExecution) {
+        return cloudletExecution.getCloudlet();
+    }
+
+    private void resetCloudletStatus(final Cloudlet cloudlet) {
+        cloudlet.reset();
+    }
+
+    private Cloudlet extractResetCloudlet(CloudletExecution cloudletExecution) {
+        Cloudlet cloudlet = extractCloudlet(cloudletExecution);
+        this.resetCloudletStatus(cloudlet);
+        return cloudlet;
+    }
+
+    private List<Cloudlet> resetCloudlets(List<CloudletExecution> toReset) {
+        return toReset
+            .stream()
+            .map(this::extractResetCloudlet)
+            .collect(Collectors.toList());
+    }
+
+    private List<Cloudlet> resetCloudletsFromVm(final Vm vm) {
+        final List<Cloudlet> cloudletsToReschedule = new ArrayList<>();
+        final CloudletScheduler cloudletScheduler = vm.getCloudletScheduler();
+
+        cloudletsToReschedule.addAll(resetCloudlets(cloudletScheduler.getCloudletWaitingList()));
+        cloudletsToReschedule.addAll(resetCloudlets(cloudletScheduler.getCloudletExecList()));
+
+        cloudletScheduler.clear();
+
+        return cloudletsToReschedule;
+    }
+
+
+    public List<Cloudlet> requestVmDestruction(final Vm vm) {
+        if(getVmExecList().contains(vm)) {
+            final List<Cloudlet> cloudletsFromVm = resetCloudletsFromVm(vm);
+            final List<Cloudlet> cloudlesSubmitted = new ArrayList<>();
+
+            for(Cloudlet cl : cloudletSubmittedList) {
+                if(cl.getVm().equals(vm)) {
+                    cloudlesSubmitted.add(cl);
+                }
+            }
+
+            cloudletSubmittedList.removeAll(cloudlesSubmitted);
+            cloudlesSubmitted.stream().forEach(this::resetCloudletStatus);
```

Comment:
Actually a question... I just realized one thing: do I actually need the `resetCloudletsFromVm` method at all? If I understand correctly all cloudlets (even those which are executing) would be on the `cloudletSubmittedList` list? I guess I can remove `resetCloudletsFromVm` altogether?

Reply:
I did this refactoring - have a look and let me know what you think.

Reply:
Yes, all cloudlets will be on `cloudletSubmittedList`. However, you can't use it because when you request the VM destruction, only running and waiting Cloudlets should be reset. Cloudlets already finished should not be touched.

Reply:
In the more recent version of code there is a test to check whether the cloudlets have been finished already: so any finished cloudlet should not be touched.